src/lib/teacher-network.ts

```diff
@@ -0,0 +1,126 @@
+import { Optional } from "utility-types";
+import { UserModelType } from "../models/stores/user";
+import { Firestore, isFirestorePermissionsError } from "./firestore";
+import { ClassDocument, OfferingDocument } from "./firestore-schema";
+import { IPortalClassInfo } from "./portal-types";
+
+export type ClassWithoutTeachers = Optional<ClassDocument, "teachers">;
+export type OfferingWithoutTeachers = Optional<OfferingDocument, "teachers">;
+
+let classTeachersPromises: Record<string, Promise<string[] | undefined>> = {};
+
+// primarily for testing
+export function clearTeachersPromises() {
+  classTeachersPromises = {};
+}
+
+// returns a promise that resolves to the list of teacher user ids (or undefined on error)
+// uses the classTeachersPromises map to guarantee that each class is only requested once
+function getClassTeachers(classUri: string, rawPortalJWT: string) {
```

Comment:
Should the functions in this file have type information declared (using TypeScript) for their return types?

Reply:
Another matter of opinion. I would say that for public APIs it makes sense to have an explicit return type, but for internal private functions it can be fine to let the TypeScript compiler infer the return type, in which case it's generally available via hover in TypeScript-aware editor. I will also sometimes add them to functions when I want the TypeScript compiler to verify that the code is actually returning what I intend for it to return, but for other functions it's not clear that there's a lot of benefit.